### ISSUE TITLE
RestAssured hangs default Restlet connector

### ISSUE THREAD

**Issue by johanhaleby** [jayway/rest-assured#322]
_From [kmin...@gmail.com](https://code.google.com/u/107309578465571932208/) on May 01, 2012 21:18:35_

What steps will reproduce the problem? 1. Extract attached zip.
2. Execute mvn clean install. What is the expected output? What do you see instead? All tests should pass but instead see the errors below from Restlet that indicate all of its worker threads are consumed and the test hangs.  Note that the tests are single threaded.

May 1, 2012 3:06:22 PM org.restlet.engine.http.connector.BaseHelper$1 rejectedExecution
WARNING: Unable to run the following server-side task: Read connection messages: true
May 1, 2012 3:06:22 PM org.restlet.engine.http.connector.BaseHelper$1 rejectedExecution
INFO: Worker service state: Normal
May 1, 2012 3:06:22 PM org.restlet.engine.http.connector.BaseHelper$1 rejectedExecution
INFO: Worker service tasks: 0 queued, 9 active, 10 completed, 19 scheduled.
May 1, 2012 3:06:22 PM org.restlet.engine.http.connector.BaseHelper$1 rejectedExecution
INFO: Worker service thread pool: 1 core size, 10 largest size, 10 maximum size, 10 current size
May 1, 2012 3:06:22 PM org.restlet.engine.http.connector.Controller run
INFO: Stop accepting new connections and transactions. Consider increasing the maximum number of threads. What version of the product are you using? On what operating system? RestAssured: 1.6.2-SNAPSHOT
Restlet: 2.0.13
Jetty: 6.1.26
MacOS: 10.7.3
Java: 1.6.0_31 Please provide any additional information below. The attached tests include multiple clients running against multiple servers.
HttpClient -> Jetty (OK)
HttpClient -> Restlet (OK)
RestAssured -> Jetty (OK)
RestAssured -> Restlet (HANG)

The meat of the tests are in test.HangTestBase which is extended in HangJettyTest and HangRestletTest.  Note I also tried a version of the test (testForHangWithRestAssuredGetManualClose) where I tried to manually close the response body stream and that fails too.

**Attachment:** [restlet-hang-v2.zip](http://code.google.com/p/rest-assured/issues/detail?id=168)

_Original issue: http://code.google.com/p/rest-assured/issues/detail?id=168_

**Comment by johanhaleby** [comment]
_From [johan.ha...@gmail.com](https://code.google.com/u/105676376875942041029/) on May 02, 2012 05:11:10_

Hi, 

Thanks for your example project and good description of the project. I've fixed the problem in trunk now for all your examples. Please try it out. I've deployed version 1.6.2-SNAPSHOT to sonatype so add:

\<repositories>
        \<repository>
            \<id>sonatype</id>
            \<url> https://oss.sonatype.org/content/repositories/snapshots/</url> \<snapshots />
        \</repository>
\</repositories>

to try it out.

**Status:** Fixed

**Comment by johanhaleby** [comment]
_From [kmin...@gmail.com](https://code.google.com/u/107309578465571932208/) on May 02, 2012 06:32:27_

Hi,
Works for me too!  Thanks again!  Would you mind briefly describing what
the issue was so that I don't make the same mistake somewhere down the
line.
Kevin.

**Comment by johanhaleby** [comment]
_From [johan.ha...@gmail.com](https://code.google.com/u/105676376875942041029/) on May 02, 2012 06:53:30_

REST Assured didn't close the idling connections and Restlet was configured to accept at most 10 connections at the same time and thus blocked subsequent requests. So I had to close the idling connections in REST Assured just as you did when using plain http client. Another problem was when getting the response body as an inputstream because then the connection cannot be closed until the user calls close on the inputstream. Thus I had to wrap the inputstream from the http client in my own inputstream that closed the connection when closing the inputstream.
